File: sdk/unity/sdk/Assets/KingstonDynamics/DragonSong/Client/DragonSong.cs
//
// This class provides a high level implementation of the DragonSong Services
//

using System;
using System.Collections;
using KingstonDynamics.DragonSong.Protocol.Messaging;
using KingstonDynamics.DragonSong.Protocol.Messaging.Type.Authentication;
using KingstonDynamics.DragonSong.Protocol.Messaging.Type.Heartbeat;

namespace KingstonDynamics.DragonSong.Client
{
    public class DragonSong
    {
        //Data Connection
        private readonly DragonNet _network;
        
        //Packet Queue
        private readonly Queue _packets;

        public DragonSong(DragonConfig config)
        {
            _network = new DragonNet(config);
            _packets = new Queue();

            if (config.AutoConnect)
            {
                _network.Connect();
            }
        }

        public void Connect()
        {
            _network.Connect();
        }

        public void Disconnect()
        {
            _network.Disconnect();
        }

        public void SendPing()
        {
            var message = new HeartbeatPing();
            Write(message);
        }
        
        public void SendAuthHandshake()
        {
            var message = new AuthenticationLogin();
            
            throw new NotImplementedException();
        }

        public void JoinChannel(string channelID)
        {
            throw new NotImplementedException();
        }

        public void PublishMessage(string channelID, string message)
        {
            throw new NotImplementedException();
        }

        public void SendPrivateMessage(string recipientID, string message)
        {
            throw new NotImplementedException();
        }

        private void Write(Message message)
        {
            _network.Write(message);
        }
        
        //
        // HANDLE RECEIVED PACKETS
        //
        public void Process()
        {
            if (_packets.Count != 0)
            {
                _packets.Dequeue();
            }
        }
    }
}

File: src/client/src/main/java/com/kdyncs/dragonsong/client/DragonSong.java
package com.kdyncs.dragonsong.client;

import com.kingstondynamics.dragonsong.protocol.message.Message;
import com.kingstondynamics.dragonsong.protocol.message.type.authentication.AuthenticationLogin;
import com.kingstondynamics.dragonsong.protocol.message.type.channel.ChannelConnect;
import com.kingstondynamics.dragonsong.protocol.message.type.channel.ChannelTransmit;
import com.kingstondynamics.dragonsong.protocol.message.type.heartbeat.HeartbeatPing;
import com.kingstondynamics.dragonsong.protocol.message.type.user.UserTransmit;

import java.util.LinkedList;

/**
 * @author peter
 */
public class DragonSong {
    
    // Data Connection
    private final DragonNet network;
    private final DragonConfig config;
    
    public DragonSong(DragonConfig config) {
        
        // Setup Networking
        this.config = config;
        this.network = new DragonNet(config);
        
        new LinkedList<>();
    }
    
    public void connect() {
        network.connect();
    }
    
    public void disconnect() {
        network.disconnect();
    }
    
    public void sendAuthHandshake() {
        Message handshake = new AuthenticationLogin(config.getAPIKey(), config.getAppID(), config.getAppVer(), config.getConnectionID(), config.getDisplayName());
        network.write(handshake);
    }
    
    public void sendPing() {
        Message ping = new HeartbeatPing();
        network.write(ping);
    }
    
    public void joinChannel(String channelID) {
        Message join = new ChannelConnect(channelID);
        network.write(join);
    }
    
    public void publishMessage(String channelID, String message) {
        Message packet = new ChannelTransmit(channelID, message);
        network.write(packet);
    }
    
    public void directMessage(String privateId, String message) {
        Message packet = new UserTransmit(privateId, message);
        network.write(packet);
    }
}